draft-ietf-quic-tls.md

```diff
@@ -866,6 +867,81 @@ the connection that is hosted on stream 0.  This sequence number is not visible
 to QUIC.
 
 
+## Packet Number Protection {#pn-encrypt}
+
+QUIC packets are protected using a key that is derived from the current set of
+secrets.  The key derived using the "pn" label is used to protect the packet
+number from casual observation.  The packet number protection algorithm depends
+on the negotiated AEAD.
+
+Packet number protection is applied after packet protection is applied (see
+{{aead}}).  The ciphertext of the packet is sampled and used as input to an
+encryption algorithm.  For packets with a long header, the ciphertext starting
+immediately after the packet number is used (that is, octet 17 onwards).  For
+packets with a short header, the packet number length is not known, so it is
+assumed to be its largest possible length (4 octets).  Thus, for a short header,
+the sampled ciphertext starts at either octet 5 when the connection ID is
+omitted, or octet 13 when the connection is present.
+
+The protected packet might not have produced enough input for the negotiated
+packet protection algorithm.  This might happen if a packet with a short header
+contains minimal data and uses a packet number encoding that is shorter than 4
+octets.  Additional zero octets are added to the end of the sequence to reach
+the required amount of data.
```

Comment:
It might make sense to use the last N bytes of the packet as IV. 

By doing so, I think that we can remove the zero-padding rule introduced in this paragraph, since it is guaranteed (at least for the algorithms being defined here) that the length of the AEAD tag is longer than that required for the IV.

Reply:
I considered this, and it does have that advantage.  However, I have been told that reading that far ahead is problematic for some implementations, performance-wise.

The odds of actually having a short packet are tiny.  For a 1 octet packet number, you need to have <3 octets of payload.  For a 2 octet packet number, then you need <2 octets of payload.  And right now only PADDING is one octet.  PING will eventually go back to 1 octet as well.  That's for a 2^-56 bound.  At 2^-60, STREAM, BLOCKED, and STREAM_ID_BLOCKED can be 2 octets.  MAX_DATA also, if you believe that <64 octets is ever something you would see on the wire.

Reply:
Instead of zero padding, the input could sampled from protected packet content including cipher text and, for short cipher text, also the authentication tag.

But

I don't like that it is not possible to do AEAD verification in-place without modifying the packet content. Thus AEAD should cover the encrypted packet number, not the plain text packet number. However, then it does not work to use the cipher nor the tag as input.

Reply:
@martinthomson Thank you for elaborating the issue behind the design decision. That makes sense to me.

@mikkelfj
> Instead of zero padding, the input could sampled from protected packet content including cipher text and, for short cipher text, also the authentication tag.

The issue is that it is impossible to tell where the ciphertext or the authentication tag begins, assuming that the length of the packet number field will become part of the protected packet number.

Consider the case where we have protected packet number (1 octet), ciphertext (1 octet), tag (16 octets). 

The length of the packet number is encoded within the protected packet number. We need to find entropy from the octets that are guaranteed to be not part of the protected packet number.

The text in the PR proposes to take from the fifth octet of the encrypted triplet (i.e. protected packet number, ciphertext, tag). That would mean that the minimum length of the entropy would be 14 bytes.

> I don't like that it is not possible to do AEAD verification in-place without modifying the packet content.

My assumption was that we would nevertheless be required to modify the packet header before feeding it as AAD to AEAD, assuming that we would either have greasing (in the packet header) or at least one spin bit. Is that no longer the case?

Reply:
> The issue is that it is impossible to tell where the ciphertext or the authentication tag begins

I see. But I don't think it would be a big deal to expose the length bits in clear text. But then again, zero padding is also fine.

> My assumption was that we would nevertheless be required to modify the packet header 

I would assume you could apply AEAD after spin bits etc.. However, packet numbers are different because they are the IV for the AEAD itself and because their encryption is derived from the same AEAD.

Reply:
See my "barrier" comment above. The only requirement is that one does not re-encrypt the bytes used for the IV, because that would lead to trouble. So, set a barrier before these bytes, and never XOR bytes after that barrier.

Reply:
I'm not sure what your proposal is there.  What I have described takes at least 14 octets always.  I think that's enough, even with the birthday bound on 112 bits being pretty low: with AES you have to re-key long before you hit that many packets, and I'd say that you'd want to do the same with ChaCha even if it remains cryptographically sound for far longer.

Reply:
What I have in mind is the following:

1) Assume that after encryption but before packet number encryption the packet is composed of encrypted payload (ccc...ccc), packet number (pp), and other header bytes (h)
~~~
    hhhhhhppppccccccc...ccc
~~~
After packet number encryption the "p" bytes will be encrypted (P). Some of the leading bytes in the ciphertext may also be re-encrypted (C):
~~~
    hhhhhhPPPPCCCCccc...ccc
~~~
This only works if the receiver can find out whether the "cipher text input" starts, without knowing the exact length of the packet encoding. At that point it only knows the position of the first packet number byte, and the length of the packet. I would like a way to specify the length of the "PN encryption" and the offset of the ciphertext sample without requiring the receiver to exactly know the packet number length.

My proposition is to set the "offset" value to MIN(pn_start + 8, packet_length-16), and to encrypt the bytes between "pn_start" and "offset".

Reply:
I was about to propose something similar to @huitema but then thought zero padding might do just as well.

However, I think it is better to take the tail content when the default start overreach because then you can always consume the content directly without moving it to a separate buffer. It's a quite simple solution.